### PR TITLE
FINERACT-803: validate username uniqueness

### PR DESCRIPTION
**Pull request by jamesidw**
## Description
Adds validation for uniqueness to the submitted username when creating/updating a user. Non-unique username was causing abrupt crash [see https://issues.apache.org/jira/browse/FINERACT-803?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel]

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Commit message starts with the issue number from https://issues.apache.org/jira/projects/FINERACT/. Ex: FINERACT-646 Pockets API.

- [x] Coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions have been followed.

- [x] API documentation at https://github.com/apache/fineract/blob/develop/api-docs/apiLive.htm has been updated with details of any API changes.

- [x] Integration tests have been created/updated for verifying the changes made.

- [x] All Integrations tests are passing with the new commits.

- [x] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the list.)

Our guidelines for code reviews is at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide

### CODE REVIEW

**Comment by santoshconflux** [comment]
@jamesidw , Thanks for the PR.  @awasum or @vorburger , Could you kindly review this PR?

File: fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java
Comment (awasum) [inline fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java]: This assumes the username will always be updated and will be different from the current one. In real life, the username usually stays the same when updating a user. Do the validation just for creation first and lets see if this resolves: https://issues.apache.org/jira/browse/FINERACT-803 as the issue occurs during creation.

**Review by awasum — CHANGES_REQUESTED** [review]

File: fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java
Comment (jamesidw) [reply fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java]: @awasum I'd changed the approach so we ignore the repeated username if it is the same as the current one but still flag the validation failure if the request would genuinely violate the unique constraint.

Let me just drop the validation in the update function as you've suggested.

**Comment by awasum** [comment]
@santoshconflux  this requires functional testing. Can you pull this and test? TEst the edit user (change the username of an existing user) and let us know.

**Review by vorburger — APPROVED** [review]
LGTM, but I'll let @awasum merge it after @santoshconflux tested... 

@jamesidw in an ideal world you'd also contribute an integration test case for this, "proving" that changing existing users still works, and that new users with duplicate usernames cannot be created... is that something you'd be interested in contributing?

**Comment by jamesidw** [comment]
@vorburger Sure thing. I'll give that a shot and update the PR.

**Comment by santoshconflux** [comment]
@awasum , @vorburger , I tested locally pulling the code: 
We are getting the validation like this in create self service user page :
create self service user

Can  we get the meaningful validation like this which is shown while creating user by inputting duplicate username(Admin>Users>Create User or edit existing user with duplicate username)

edit user in Users

**Comment by awasum** [comment]
> @awasum , @vorburger , I tested locally pulling the code:
> We are getting the validation like this in create self service user page :
> create self service user
> 
> Can we get the meaningful validation like this which is shown while creating user by inputting duplicate username(Admin>Users>Create User or edit existing user with duplicate username)
> 
> edit user in Users

@jamesidw  would you like to also look at this?

**Comment by santoshconflux** [comment]
@jamesidw , In the ticket https://issues.apache.org/jira/browse/FINERACT-803, I mentioned that the validation should be "Username is already taken". That is also good. 
But let it be "User with username 'username' already exists" as in screenshot,  in order to maintain the uniformity.

**Review by awasum — APPROVED** [review]

**Comment by jamesidw** [comment]
@awasum The error message should now match what @santoshconflux shared. 

I have a question though: what is being displayed in the screenshots is actually `userMessageGlobalisationCode` but it should be `defaultUserMessage` (at least that's what the other error messages seem to use). Is that a mistake on the frontend?

File: fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java
Comment (mohitsinha) [inline fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java]: I believe there is no point going inside this code block, if the username is blank/empty.
```suggestion
        if (StringUtils.isEmpty(username)) {
```
```suggestion
        if (StringUtils.isBlank(username)) {
```
@vorburger  @awasum What's your opinion on this?

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/useradministration/users/UserHelper.java
Comment (mohitsinha) [inline fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/useradministration/users/UserHelper.java]: Can we remove these print statements?

**Review by mohitsinha — CHANGES_REQUESTED** [review]
Looks good to me. 
Some minor suggestions from my side.

File: fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java
Comment (jamesidw) [reply fineract-provider/src/main/java/org/apache/fineract/useradministration/service/UserDataValidator.java]: That makes sense. I've applied the optimization suggestion @mohitsinha

File: fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/useradministration/users/UserHelper.java
Comment (jamesidw) [reply fineract-provider/src/integrationTest/java/org/apache/fineract/integrationtests/useradministration/users/UserHelper.java]: Sure. I've removed them

**Comment by vorburger** [comment]
I was about to merge this one, but then noticed that you @mohitsinha now also are a commiter, or are just about to become one?  Let's use this PR as the first opportunity for you to actually merge a contribution from someone yourself! @awasum I'll hold off from merging this to give @mohitsinha a chance to do this, perhaps you'd like to do the same.

**Comment by awasum** [comment]
> I was about to merge this one, but then noticed that you @mohitsinha now also are a commiter, or are just about to become one? Let's use this PR as the first opportunity for you to actually merge a contribution from someone yourself! @awasum I'll hold off from merging this to give @mohitsinha a chance to do this, perhaps you'd like to do the same.

Yes. @mohitsinha go ahead, reviews this and merge.

**Review by mohitsinha — APPROVED** [review]

**Comment by mohitsinha** [comment]
I will merge this, once I am done with the onboarding process

**Comment by mohitsinha** [comment]
@jamesidw Can you combine all the 7 commits into 1?

**Comment by jamesidw** [comment]
@mohitsinha I've squashed the commits

**Comment by santoshconflux** [comment]
@mohitsinha , @jamesidw , @awasum . I just fetched the latest code to locally from develop branch. I am still getting the error: 
 "field is required
validation.msg.user.username.constraint.violation.duplicate.username -" 
Tested in incognito mode as well, but getting the above error.

**Comment by mohitsinha** [comment]
Should we revert this change

**Comment by jamesidw** [comment]
@santoshconflux please share the frontend repository you are using, I'd like to see what field is being displayed from the response

**Comment by santoshconflux** [comment]
@jamesidw , I am using this one.
https://github.com/openMF/community-app from develop branch.

**Comment by jamesidw** [comment]
Thanks @santoshconflux from this file [notificationWidget.js](https://github.com/openMF/community-app/blob/develop/app/scripts/modules/notificationWidget.js) it seems the correct approach is to use the `userMessageGlobalisationCode` instead. There is [already a code](https://github.com/openMF/community-app/blob/c48f89dc3397aaa16cad1cabfd6324e802e12fcd/app/global-translations/locale-en.json#L2372) that can work for this on the frontend so the fix would entail changing the `userMessageGlobalisationCode` being returned to match that.

@mohitsinha not sure what the correct approach is:

* Revert the changes and create a new PR?
* Create a new PR to fix what's already been merged?

Please advise

**Comment by mohitsinha** [comment]
@jamesidw Let's first find the issue. 
Reverting the change should be more advisable as other users won't be affected.

I don't really have a problem if we add a new PR on top of the existing changes. If the bug is minor and won't affect other users, we can consider this option.

**Comment by jamesidw** [comment]
@mohitsinha The issue as I understand it is: 

- The frontend uses the `userMessageGlobalisationCode` to create the message to be displayed. 
- I created a new `userMessageGlobalisationCode` in this PR and the frontend doesn't recognise it so it displays the `userMessageGlobalisationCode` text directly.

While looking through the frontend repository that @santoshconflux shared, I realised that the message he asked me to create already existed on the frontend...which meant it already existed on the backend as well. Indeed [this function](https://github.com/apache/fineract/blob/fb1e3d7c18fdcf208654f974dcb16f3eb34dad33/fineract-provider/src/main/java/org/apache/fineract/useradministration/service/AppUserWritePlatformServiceJpaRepositoryImpl.java#L343) was supposed to handle the exception and set the correct error response.

Let's revert the changes

**Comment by mohitsinha** [comment]
@awasum @vorburger What's the standard way to revert a commit. 
Do I need to create a new PR after reverting the commit?
Or can this be reverted directly?

**Comment by awasum** [comment]
> @awasum @vorburger What's the standard way to revert a commit.
> Do I need to create a new PR after reverting the commit?
> Or can this be reverted directly?

There should be a direct link or button somewhere on github to revert the PR.

**Comment by awasum** [comment]
There is a revert button or link where you merged the PR...See
image

https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/reverting-a-pull-request

@santoshconflux  should describe exactly the test criteria so we all no when this will pass QA. Looks to me like the problem is understood but the QA criteria for a pass is not... I dont know if am making sense to you all.

**Pull request merged**
